streams/src/main/java/org/apache/kafka/streams/kstream/Suppressed.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl.IntermediateSuppressionImpl;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import java.time.Duration;
+
+@SuppressWarnings("unused") // complaint about the top-level generic parameters, which actually are necessary
+public interface Suppressed<K, V> {
+    interface TimeDefinition<K, V> {
+        long time(ProcessorContext context, K k, V v);
+    }
+
+    interface EagerBufferConfig<K, V> extends BufferConfig<K, V, EagerBufferConfig<K, V>> {
+
+    }
+
+    interface StrictBufferConfig<K, V> extends BufferConfig<K, V, StrictBufferConfig<K, V>> {
+
+    }
+
+    interface BufferConfig<K, V, BC extends BufferConfig<K, V, BC>> {
+        static <K, V> BufferConfig<K, V, ? extends BufferConfig> serializedWith(final Serialized<K, V> serialized) {
+            return new SuppressedImpl.EagerBCImpl<>(serialized);
+        }
+
+        BC serialize(final Serialized<K, V> serialized);
+
+        static <K, V> EagerBufferConfig<K, V> boundedByKeys(final long maxKeys) {
+            return new SuppressedImpl.EagerBCImpl<>(maxKeys);
+        }
+
+        BC bufferKeys(final long maxKeys);
+
+        static <K, V> EagerBufferConfig<K, V> boundedBySize(final long maxBytes) {
+            return new SuppressedImpl.EagerBCImpl<>(maxBytes);
+        }
+
+        BC bufferBytes(final long maxBytes);
+
+        static <K, V> StrictBufferConfig<K, V> unbounded() {
+            return new SuppressedImpl.StrictBCImpl<>();
+        }
+
+        StrictBufferConfig<K, V> unboundedBuffer();
+
+        StrictBufferConfig<K, V> spillToDiskWhenFull();
+
+        StrictBufferConfig<K, V> shutDownWhenFull();
+
+        EagerBufferConfig<K, V> emitEarlyWhenFull();
```

Comment:
Added these methods in lieu of the BufferFullStrategy enum

Reply:
This allows us to insist on a `StrictBufferConfig` for the final results use case.

Reply:
see: https://github.com/apache/kafka/pull/5567/files#r215067976